packages/astro/src/core/endpoint/index.ts

```diff
@@ -59,8 +58,18 @@ export function createAPIContext({
 			});
 		},
 		ResponseWithEncoding,
-		preferredLocale: preferredLocale,
-		preferredLocaleList: preferredLocaleList,
+		get preferredLocale(): string | undefined {
+			if (locales) {
+				return computePreferredLocale(request, locales);
+			}
+			return undefined;
+		},
+		get preferredLocaleList(): string[] | undefined {
+			if (locales) {
+				return computePreferredLocaleList(request, locales);
+			}
+			return undefined;
+		},
```

Comment:
Should we also handle caching for the endpoint API context. However it doesn't seem like `cookies` is also cached here. Not a blocker for the PR for me though.